scandium-core/src/main/java/org/eclipse/californium/scandium/dtls/Connection.java

```diff
@@ -283,7 +292,62 @@ public boolean addInbound(Record record) {
 	 * 
 	 * @return the next record or {@code null} if the buffer is empty.
 	 */
-	public Record getNextInboundRecord() {
-		return incoming.poll();
+	private Record getNextInboundRecord() {
+
+		synchronized(incoming) {
+			return incoming.poll();
+		}
+	}
+
+	/**
+	 * Triggers processing of the next record from this connector's buffer.
+	 * <p>
+	 * This method schedules a new task on the executor which
+	 * <ol>
+	 * <li>takes the next record from this connection's buffer (if available)</li>
+	 * <li>invokes the processor's {@linkplain RecordProcessor#process(Record, Connection) process} method</li>
+	 * <li>invokes this method</li>
+	 * </ol>
+	 * <p>
+	 * This method uses an atomic boolean to make sure that only one record from this connection's buffer
+	 * is processed at a time, thus making sure that all records are processed in order of reception.
+	 * 
+	 * @param processor The object that should be used for processing the records.
+	 * @param executor The service to use for scheduling the processing tasks.
+	 */
+	public void processNextMessage(final RecordProcessor processor, final ExecutorService executor) {
```

Comment:
There is a lot of synchronization tools (blocking queue + synchronized block+ atomic boolean)

I tried to find a better way :
- First time by removing synchronized block but I can not find any thread safe way to do.
- Then, I tried to just use synchronize block, we don't really need atomic boolean and concurrent queue, but I can not find a non concurrent queue implementation with a maximum size, so I'm just able to remove the atomic boolean...
- So finally, I think it's ok like that.

Another solution could be to use a [Striped Executor Services](http://www.javaspecialists.eu/archive/Issue206.html). That way, we lost the queue size limitation. (Maybe we don't really need it)